### PR TITLE
External Test Runner Support

### LINKED ISSUE #252: Memory consumption is very high on large projects
**Describe the bug**
This is a bit vague -- sorry -- but our project at work has gotten large enough (19 projects, 107 components) that when a low-level component (or `deps.edn` file) is updated and it triggers a test run of almost all projects and all components, our CI job gets killed for memory usage (BitBucket, 2x instance, already the largest we can have).

**To Reproduce**
Very hard to reproduce!

**Expected behavior**
`poly test` should free up heap and non-heap that it doesn't use (tricky: I suspect at least some of this is the large classloader space created by so many projects and so many components, but haven't proved that yet).

**Partial Workaround**
We're running a fork of Polylith that has a call to `(System/gc)` in the test orchestrator after each test suite has run. This only signals to the JVM that we'd like garbage collection as a hint, but it does seem to have been sufficient to reduce the memory footprint _most of the time_. See https://github.com/seancorfield/polylith/compare/4e946eab081cdef317377da2e9b2f61158dbc422..as-alias 

We may continue to tweak min/max heap sizes via JVM options to see if we can get a better outcome, but we've already had to set min/max to control heap usage on other environments so we're threading a needle here already. I don't know if there's an "ideal" solution to this (but Polylith's classloader isolation is problematic in a couple of areas as noted in other discussions on Slack and other GH issues).

### LINKED ISSUE #253: Classloader isolation has "leaks"
**Describe the bug**
Code that uses global JVM thread pools (such as code based on fork/join) can end up loading classes into those threads which then conflict with those same classes loaded into the classloaders `poly test` creates to run code.

**To Reproduce**
Exceedingly hard to reproduce(!) and even in some repro cases, it is intermittent. We've managed to mitigate _most_ instances by moving `reify` calls into the main thread where possible in our code (so no new class instances get loaded in the fork/join thread classloader) but there are places where we just can't do this.

**Expected behavior**
Given the use of isolated classloaders, this _is_ the expected behavior -- it's just problematic. Running tests in a subprocess would resolve this (and other classloader issues) but would make Polylith's test runner logic much more complex I suspect.

**Partial Workaround**
In addition to reorganizing our code as described above, we have found that changing the `*use-context-classloader*` binding to `false` seems to help resolve more cases of this issue. The default is `true` so I'm not sure why that `binding` was ever introduced, but at least it provides for a smaller "fix": https://github.com/seancorfield/polylith/commit/5945bd2cf5b36f07a759744f1de702c79970a9b4

Since we made that change on my fork, we have not run into the ClassCastException that the cross-classloader issue described above was causing -- so I think it's a safe change to make and will help others who might be using fork/join code.

### LINKED ISSUE #260: External process test runner
**Is your feature request related to a problem? Please describe.**
Polylith's classloader isolation approach is problematic for a number of reasons (issues and PRs have already been submitted/merged to mitigate some of these problems, including excessive heap usage on very large projects). Running each project's test suite in an isolated _subprocess_ would address all of these issues, once and for all.

**Describe the solution you'd like**
A new test runner, that `poly test` can invoke as a subprocess, to run a project's test suite in complete isolation. This requires some minor modifications to Polylith to support this:
* Instead of running the setup/teardown for a project once and then running multiple test runners on the project, run the setup/teardown for each test runner -- this is a prerequisite for an external process since the setup/teardown must run in the same context as the tests, i.e., in the subprocess.
* Modify the (currently unconditional) construction of the classloader such that it can be deferred until a test runner seems to need it -- also a prerequisite for an external process since it wouldn't need the classloader (only the classpath) and repeatedly building the classloader is part of what causes the excessive heap issues in very large projects.
* Provide a way for a test runner to indicate that it wishes to be run as a subprocess via a classpath: it would also be responsible for running the setup and teardown functions.
* Design a _simple_ command-line API for such test runners -- proposal: pass the color-mode string, the project-name, the setup-fn name if present, the list of namespaces to require and run tests from, and the teardown-fn name if present.

**Describe alternatives you've considered**
Attempts to mitigate classloader isolation and memory usage have already been applied and are insufficient.

**Additional context**
We are using my fork of Polylith with an external test runner at work and it addresses all our issues and seems to otherwise function identically to the current in-process test runner. A PR for a cleaned-up version of this will be forthcoming.

### PR DESCRIPTION
**Pull request by furkan3ayraktar**
Another stab at the issue #260 

This PR aimed to develop a cleaner solution with fewer changes to the existing code. It is a result of my discussion with @tengstrand and is heavily inspired by @seancorfield's PR #262. 

See [seancorfield/polylith-external-test-runner](https://github.com/seancorfield/polylith-external-test-runner) as an example of an external test runner.

There is no documentation about custom test runners, except the doc string in the `test-runner-contract` interface. It would be nice to add a new page to the poly documentation and explain how to develop custom in-process and external test runners.

Resolves #260, resolves #252, resolves #253
Also closes #262

### CODE REVIEW

**Comment by seancorfield** [comment]
That does look cleaner and more "idiomatic" in keeping with how the existing test runner is constructed.

Have you tested it works with my external test runner as-is? (curious as to whether it would cause any changes in my project; I can test this out tomorrow at work)

**Comment by furkan3ayraktar** [comment]
I've run the Polylith project's test using your external test runner, and it worked as-is. I did not intentionally change any input that goes to the test runner, so I assume it should work with your project at work. Let me know if I caused any issues.

**Comment by furkan3ayraktar** [comment]
I've found two issues with the test runner and addressed them with two commits.

**Issue 1**
If a project does not exist in the `workspace.edn` file, the tests did not run for that project even though its name was logged during the test run. The reason is that the project missing from the `workspace.edn` file did not get a default test runner in its settings; hence, no tests ran.

[This commit](https://github.com/polyfy/polylith/pull/263/commits/0628224b4e29671f39fc41463b377675e2a5b63b) fixes the issue.

**Issue 2**
The `run` function in the `test-runner-orchestrator` component stopped executing tests in other projects if the `run-tests-for-project` returned a falsy value for a project. The `run-tests-for-project` returns `nil` if there are no test runners in the project settings that see test sources. Hence it returns `nil` if there are no test runners for the project. I noticed this connected to issue 1. Even though fixing issue 1 ensures there will always be a test runner for a project, it is still the correct behavior to execute tests for projects with test runners seeing test sources even though one of the projects doesn't run any tests.

[This commit](https://github.com/polyfy/polylith/pull/263/commits/f0a93b529a60f1376c049becb3a17dce0c6a0be6) fixes the issue.

File: components/test-runner-orchestrator/src/polylith/clj/core/test_runner_orchestrator/core.clj [hunk 1]
```diff
@@ -56,19 +56,48 @@
                           " project: " e))
             (throw e))))))
 
+(defn execute-setup-fn [project color-mode {:keys [setup-fn runner-ns class-loader]}]
```
Comment (seancorfield) [inline components/test-runner-orchestrator/src/polylith/clj/core/test_runner_orchestrator/core.clj]: You have `runner-ns` in both `execute-*-fn` functions, but the key is `:process-ns` in `run-tests-for-project`, so this attempts to run the setup function without a class loader and blows up.

**Comment by seancorfield** [comment]
Per the comment above, this breaks at work because we use `:setup-fn`/`:teardown-fn` in all of our projects and I'm guessing their are no tests around that for the external test runner. When that's fixed, I'll test the PR again at work.

File: components/test-runner-orchestrator/src/polylith/clj/core/test_runner_orchestrator/core.clj (hunk 1 above)
Comment (furkan3ayraktar) [reply components/test-runner-orchestrator/src/polylith/clj/core/test_runner_orchestrator/core.clj]: Ah, right! I shouldn’t commit at 1 am. I’ll fix it and add tests for those functions. I’ll let you know. Thanks for testing it out.

Comment (furkan3ayraktar) [reply components/test-runner-orchestrator/src/polylith/clj/core/test_runner_orchestrator/core.clj]: I've fixed this here: https://github.com/polyfy/polylith/pull/263/commits/010d5e853247cdaf0d9e6c6079b5aca64a86db98

**Comment by furkan3ayraktar** [comment]
I've fixed the issue in the comment above. Let me know if you find further problems.

**Comment by seancorfield** [comment]
That seems to work now! Thank you! I'll switch us over to this at work "full-time" until it is merged and let you know if any other issues come up.

File: components/workspace/src/polylith/clj/core/workspace/settings.clj [hunk 2]
```diff
@@ -8,12 +8,24 @@
 (defn undefined-project [index project-name]
   [project-name {:alias (str "?" (inc index))}])
 
+(def default-test-runner
```
Comment (imrekoszo) [inline components/workspace/src/polylith/clj/core/workspace/settings.clj]: There is similar code in `polylith.clj.core.workspace-clj.project-settings` for projects that are in workspace.edn. With this code here I'm wondering whether it would make sense to move all of that old code into this namespace so they are in one place?

Comment (imrekoszo) [reply components/workspace/src/polylith/clj/core/workspace/settings.clj]: Nice catch by the way, looks like I should have implemented setting the default in this file as opposed to the other namespace.

File: components/test-runner-contract/src/polylith/clj/core/test_runner_contract/interface.clj [hunk 3]
```diff
@@ -47,23 +47,33 @@
...
 
-  {:test {:create-test-runner my.namespace/create} ; to use it globally
+  {:test {:create-test-runner [my.namespace/create]} ; to use it globally
 
    :projects {; to use it only for a project
-              \"project-a\" {:test {:create-test-runner my.namespace/create}}
+              \"project-a\" {:test {:create-test-runner [my.namespace/create]}}
               ; to reset the global setting to default
-              \"project-b\" {:test {:create-test-runner :default}}}}"
+              \"project-b\" {:test {:create-test-runner [:default]}}}}"
```
Comment (imrekoszo) [inline components/test-runner-contract/src/polylith/clj/core/test_runner_contract/interface.clj]: suggestion: add another example project here showing that multiple runners can also be provided in the vector

Comment (imrekoszo) [reply components/test-runner-contract/src/polylith/clj/core/test_runner_contract/interface.clj]: The value of `:create-test-runner` doesn't have to be a vector by the way - `:default` and `[:default]` result in the same behavior, the former gets wrapped in a vector if that's how it's provided: https://github.com/polyfy/polylith/blob/abd20bb656e2cdd712732d80e1238316d3ea1b7f/components/workspace-clj/src/polylith/clj/core/workspace_clj/project_settings.clj#L18

Not sure if it's worth its own example here.

**Comment by imrekoszo** [comment]
I checked this with polylith-kaocha and found no breakages.

File: components/workspace/src/polylith/clj/core/workspace/settings.clj (hunk 2 above)
Comment (furkan3ayraktar) [reply components/workspace/src/polylith/clj/core/workspace/settings.clj]: Yeah, I noticed this when I added a new project at work but forgot to add it to the `workspace.edn` file 😂 

I didn’t know where to put the shared code; it seemed like the two places did different things, and merging them would be wrong. What do you think @tengstrand?

File: components/test-runner-orchestrator/src/polylith/clj/core/test_runner_orchestrator/core.clj [hunk 4]
```diff
@@ -108,26 +137,26 @@
...
-            runner-opts (merge opts
-                               {:class-loader class-loader
-                                :eval-in-project (->eval-in-project class-loader)})]
+            all-paths (into [] cat [(:src paths) (:test paths) lib-paths])
+            class-loader-delay (delay (common/create-class-loader all-paths color-mode))]
         (when is-verbose (println (str "# paths:\n" all-paths "\n")))
         (doseq [current-test-runner test-runners-seeing-test-sources]
-          (when-not (setup-failed?)
+          (let [process-ns (when (test-runner-verifiers/valid-external-test-runner? current-test-runner)
+                             (test-runner-contract/external-process-namespace current-test-runner))
```
Comment (imrekoszo) [inline components/test-runner-orchestrator/src/polylith/clj/core/test_runner_orchestrator/core.clj:145]: question: if I'm reading the code right, if this ends up being non-null, then we're in external mode. For a test runner wanting to support both in-process and external modes, the way to do it would be to provide separate `create` functions for the two modes, is that correct?

**Comment by imrekoszo** [comment]
Question to all of you: would it make sense to port over some of the code that deals with launching the external process from Sean's runner implementation?

Perhaps it's a little early for that. I'm thinking, it wouldn't be impossible to create a completely separate project (or even something inside the poly tool), which could implement the extra protocol added in this PR and "externalize" any in-process test runner, something along the lines of what Sean and I were [debating on slack](https://clojurians.slack.com/archives/C013B7MQHJQ/p1668195092811809?thread_ts=1668190791.646639&cid=C013B7MQHJQ).

File: components/workspace/src/polylith/clj/core/workspace/settings.clj (hunk 2 above)
Comment (furkan3ayraktar) [reply components/workspace/src/polylith/clj/core/workspace/settings.clj]: @tengstrand These are the two places in question:

[polylith.clj.core.workspace.settings](https://github.com/polyfy/polylith/blob/0628224b4e29671f39fc41463b377675e2a5b63b/components/workspace/src/polylith/clj/core/workspace/settings.clj)
[polylith.clj.core.workspace-clj.project-settings](https://github.com/polyfy/polylith/blob/0628224b4e29671f39fc41463b377675e2a5b63b/components/workspace-clj/src/polylith/clj/core/workspace_clj/project_settings.clj)

File: components/test-runner-contract/src/polylith/clj/core/test_runner_contract/interface.clj (hunk 3 above)
Comment (furkan3ayraktar) [reply components/test-runner-contract/src/polylith/clj/core/test_runner_contract/interface.clj]: @imrekoszo I've added more examples now. Let me know what you think.

File: components/test-runner-orchestrator/src/polylith/clj/core/test_runner_orchestrator/core.clj (hunk 4 above)
Comment (furkan3ayraktar) [reply components/test-runner-orchestrator/src/polylith/clj/core/test_runner_orchestrator/core.clj:145]: A test runner can either be an in-process test runner and only reify the `TestRunner` protocol or an external test runner and reify both `TestRunner` and `ExternalTestRunner` protocols. It needs to return a valid external process namespace for it to be considered an external test runner.

**Comment by furkan3ayraktar** [comment]
> Question to all: would it make sense to port over some of the code that deals with launching the external process from Sean's runner implementation?

This would remove the boilerplate needed for handling the Java subprocesses and running the setup/teardown functions. However, I don't know if it's important for an external test runner to control the process of creating the "external environment" needed to run the tests. I don't think I have enough knowledge to argue about this topic. What do you think, @seancorfield? Do you think it's valuable to control the creation of the Java subprocesses, or should we hide it inside the poly?

File: components/test-runner-contract/src/polylith/clj/core/test_runner_contract/interface.clj [hunk 5]
```diff
@@ -66,14 +66,42 @@
...
+
+  Adding a test runner definition to a project will override the global test
+  runner. The project-a will use the global test runner, `my.namespace/create`
+  whereas project-b will use the default test runner.
+
+  {:test {:create-test-runner my.namespace/create}
+
+   :projects {\"project-a\" {:alias \"a\"}
+              \"project-b\" {:alias \"b\"
+                             :test {:create-test-runner :default}}}}"
```
Comment (imrekoszo) [inline components/test-runner-contract/src/polylith/clj/core/test_runner_contract/interface.clj]: Reads very well in my opinion!

Comment (furkan3ayraktar) [reply components/test-runner-contract/src/polylith/clj/core/test_runner_contract/interface.clj]: Thank you!

**Comment by seancorfield** [comment]
> This would remove the boilerplate needed for handling the Java subprocesses and running the setup/teardown functions. However, I don't know if it's important for an external test runner to control the process of creating the "external environment" needed to run the tests. I don't think I have enough knowledge to argue about this topic. What do you think, @seancorfield? Do you think it's valuable to control the creation of the Java subprocesses, or should we hide it inside the poly?

I don't think we have enough knowledge/experience with this yet to decide to put the process spawning inside `poly` itself. As my code stands, there are several "knobs" that users might need to tweak. It doesn't support `JAVA_CMD` or a "standard" way to find the `java` command; I'm not very happy with the handling of JVM options (and that's critical to our test environment); it hasn't been tested on Windows (which is notorious for filename/path issues).

I _do_ feel that there's a fair bit of boilerplate/duplication between the, now three, different test runners in the `create` functions, so that might be an area that is worth looking at for pushing more of that into `poly`. Even there tho', there's already a TODO comment about narrowing what gets tested and it's possible that code might need to change between test runners. For example, it might be nice for us to have a test runner that is better integrated into our codebase at work and can make more assumptions about the environment in which it runs (for example, I might want to put the JVM options in an alias in `deps.edn` and read that via t.d.a when the runner is `create`'d).

There's also a use case that @imrekoszo raised in an earlier discussion: the test runner might prefer to be passed the classpath as an argument and build its own classloader for running tests (instead of running the subprocess directly via `java -cp ...`). We very specifically do not want that at work, but I can see why other test runners _might_ want to do that.

I can imagine someone deciding to build an external test runner based on Cognitect's `test-runner` which has a completely different API in terms of command-line arguments.

In all these cases, the actual subprocess is responsible for running the `setup-fn` and `teardown-fn`, regardless of what else the subprocess does, which means those need to be evaluated in the same context as the tests themselves.

TL;DR: adopt a "wait and see" approach to see what the community does with this new found superpower 😄

File: components/workspace/src/polylith/clj/core/workspace/settings.clj (hunk 2 above)
Comment (furkan3ayraktar) [reply components/workspace/src/polylith/clj/core/workspace/settings.clj]: @tengstrand and I discussed this and decided to move all test runner-related enrichments to the workspace map from the `workspace-clj` component to the `workspace` component.

You can find my changes here: https://github.com/polyfy/polylith/pull/263/commits/42d327e8211f0723d8c1f3cea3a4d03c847ccc9b

Let me know what you think!

**Review by imrekoszo — APPROVED** [review]
LGTM, thank you! polylith-kaocha tests are passing with 42d327e8211f0723d8c1f3cea3a4d03c847ccc9b

**Comment by seancorfield** [comment]
I updated our repo at work to use the latest SHA for this PR (42d327e) and it seems to work, with the external test runner.

**Comment by seancorfield** [comment]
I have updated https://github.com/seancorfield/polylith-external-test-runner to improve how it finds the `java` command to run the subprocess for the tests and also to support specifying JVM opts as an alias, which makes it a lot easier to pass multiple JVM options to the test subprocess. See the updated README, the workspace-level `deps.edn` file, and the newly-added tests.

**Comment by seancorfield** [comment]
We've been running against this fork for a week now at work and it's working well, both locally and in CI, and has allowed us to reduce our BitBucket Pipelines CI configuration from `2x` back to the default size which has halved the cost of CI for us!

This PR also fixes #252 (memory consumption in large projects) and #253 (classloader isolation leaks).

Are there any obstacles to getting this merged at this point?
